Primary will no longer always provide JSON time attestation

awwad opened the pull request:
The Primary correctly received and validated time attestations
in the format configured in tuf.conf.METADATA_FORMAT.

However, the Primary converts time attestations into a
comprehensible JSON-compatible format after receiving and
validating them.

Due to an oversight, the Primary was failing to check the
format setting and (if appropriate) convert time attestations
back into ASN.1/DER before passing them on to the Secondary,
so regardless of the setting and regardless of the format
that Timeserver sent to Primary, Secondaries were always
receiving time attestations in JSON format.

This is now fixed. Accompanying this are the changes
necessary to the demo to allow it to now transfer this
binary data appropriately (wrapping in xmlrpc Binary objects
based on format setting).

awwad committed:
Primary will no longer always provide JSON time attestation

The Primary correctly received and validated time attestations
in the format configured in tuf.conf.METADATA_FORMAT.

However, the Primary converts time attestations into a
comprehensible JSON-compatible format after receiving and
validating them.

Due to an oversight, the Primary was failing to check the
format setting and (if appropriate) convert time attestations
back into ASN.1/DER before passing them on to the Secondary,
so regardless of the setting and regardless of the format
that Timeserver sent to Primary, Secondaries were always
receiving time attestations in JSON format.

This is now fixed. Accompanying this are the changes
necessary to the demo to allow it to now transfer this
binary data appropriately (wrapping in xmlrpc Binary objects
based on format setting).

vladimir-v-diaz commented on demo/demo_primary.py:
```diff
@@ -688,6 +688,10 @@ def get_time_attestation_for_ecu(ecu_serial):
+    # If we're using ASN.1/DER, then the attestation is binary data we're about
+    # to transmit via XMLRPC, so we should wrap it appropriately:
+    attestation = xmlrpc_client.Binary(
+        primary_ecu.get_last_timeserver_attestation())
```
The comment and code don't seem to match.

The code always seems to send the `attestation` in Binary, but the comment says that "If we're using ASN.1/DER, the attestation is wrapped in binary data..."

I was expecting the attestation to be wrapped in Binary if tuf.conf.METADATA_FORMAT == 'der'.

awwad replied:
Good catch, thanks.

vladimir-v-diaz commented on uptane/clients/primary.py:
```diff
@@ -806,8 +806,26 @@ def get_last_timeserver_attestation(self):
     else:
-      return self.all_valid_timeserver_attestations[-1]
+      raise uptane.Error('Unable to convert time attestation as configured. '
+          'Unexpected format setting: ' + repr(tuf.conf.METADATA_FORMAT))
```
Perhaps we should list the expected formats.

awwad committed:
Bugfix to previous commit: wrap as binary object only as necessary

Condition use of Binary object for XMLRPC transfer in the demo
code on whether or not we're using ASN.1/DER format (as intended).

Thanks for catching this, @vladimir-v-diaz.

awwad commented:
Addressed.
